### DISCUSSION TITLE
Derivations from Classes

### DISCUSSION THREAD

**Discussion by mlutze** [flix/flix#5768] [Ideas]
### The Problem
We want to be able to support implication relationships among classes. We already support this in some sense with class constraints (superclasses):
```scala
class Order[a] with Eq[a] { ... }
```
But this requires anyone who wants to implement `Order` also to implement `Eq`. It's not too much work in this case, but `Order` implies other classes, such as `PartialOrder`. It may quickly become a burden to implement all these classes every time.

### The Nonworking Solution
The straightforward (but system-breaking solution) is to create a generic `PartialOrder` instance with an `Order` type constraint:
```scala
instance PartialOrder[a] with Order[a] {
  def partialCompare(x: a, y: a): PartialComparison = Order.compare(x, y) |> toPartialComparison
}
```
This is illegal because `a` is a _complex instance type_: every instance must be a type constructor applied to 0 or more distinct type variables. This rule prevents overlapping instances and to a certain extent keeps overly-general instances from blocking specific instances (e.g. in this case no one could write their own `PartialOrder` instance because `a` unifies with every type).

### A Potential Solution
Rather than one general instance automatically used, we allow users to opt-in to the use of an instance.
```scala
// define how to derive PartialOrder from Order
derivation PartialOrder[a] from Order[a] {
  def partialCompare(x: a, y: a): PartialComparison = Order.compare(x, y) |> toPartialComparison
}

// use the derivation to get a free PartialOrder instance for Int
derive PartialOrder[Int] from Order[Int]

// or maybe even allow derivation at the instance declaration
instance Order[Int] deriving PartialOrder { ... }
```

This means that adding a `PartialOrder` instance for most types is a one-line change, and is especially easy at the construction of the `Order` instance.

We'll need a few extra rules to be sure there's no cyclicity. E.g. `Order` has the `Eq` constraint, therefore `Eq` cannot be derived from `Order`.

### Discussion
* Syntax
* Named derivations (maybe we could derive in two different ways from the same base class)

**Comment by mlutze** [comment]
@magnus-madsen thoughts?

**Comment by magnus-madsen** [comment]
I am still thinking about it. I am concerned about the added complexity for little gain. 

Yes, its a problem for Eq, Order, etc. Also for Functor, Applicative, and Monad.

I am just not sure if this is the right solution.
